# Data Preparation

In this notebook, the dataset is cleaned and prepared for machine learning.

In [1]:
import pandas as pd

df = pd.read_csv("../data/raw/278k_song_labelled.csv")

In [2]:
df = df.drop(columns=["Unnamed: 0"])

df.head()

,duration (ms),danceability,energy,loudness,speechiness,acousticness,instrumentalness,liveness,valence,tempo,spec_rate,labels
0,195000.0,0.611,0.614,-8.815,0.0672,0.0169,0.000794,0.7530,0.520,128.050,3.446154e-07,2
1,194641.0,0.638,0.781,-6.848,0.0285,0.0118,0.009530,0.3490,0.250,122.985,1.464234e-07,1
2,217573.0,0.560,0.810,-8.029,0.0872,0.0071,0.000008,0.2410,0.247,170.044,4.007850e-07,1
3,443478.0,0.525,0.699,-4.571,0.0353,0.0178,0.000088,0.0888,0.199,92.011,7.959809e-08,0
4,225862.0,0.367,0.771,-5.863,0.1060,0.3650,0.000001,0.0965,0.163,115.917,4.693131e-07,1


In [3]:
df.isnull().sum()

duration (ms)       0
danceability        0
energy              0
loudness            0
speechiness         0
acousticness        0
instrumentalness    0
liveness            0
valence             0
tempo               0
spec_rate           0
labels              0
dtype: int64

### Missing values

The dataset contains no missing values, so no imputation or removal of rows is required.

---

In [4]:
df.duplicated().sum()

np.int64(1678)

In [5]:
df[df.duplicated(keep=False)].sort_values(
    by=["danceability", "energy", "valence"]
).head(10)

,duration (ms),danceability,energy,loudness,speechiness,acousticness,instrumentalness,liveness,valence,tempo,spec_rate,labels
148076,259292.0,0.0662,0.00741,-36.066,0.0434,0.957,0.977,0.120,0.0388,62.310,1.673789e-07,3
148077,259292.0,0.0662,0.00741,-36.066,0.0434,0.957,0.977,0.120,0.0388,62.310,1.673789e-07,3
95619,266920.0,0.0693,0.06270,-31.924,0.0482,0.974,0.762,0.173,0.0395,78.063,1.805785e-07,3
95646,266920.0,0.0693,0.06270,-31.924,0.0482,0.974,0.762,0.173,0.0395,78.063,1.805785e-07,3
148227,177391.0,0.0699,0.11600,-21.819,0.0460,0.840,0.917,0.110,0.0361,181.307,2.593142e-07,3
238133,177391.0,0.0699,0.11600,-21.819,0.0460,0.840,0.917,0.110,0.0361,181.307,2.593142e-07,3
182280,181107.0,0.0701,0.01320,-28.395,0.0404,0.980,0.634,0.072,0.0340,75.221,2.230725e-07,3
201104,181107.0,0.0701,0.01320,-28.395,0.0404,0.980,0.634,0.072,0.0340,75.221,2.230725e-07,3
204676,329907.0,0.0704,0.17900,-17.395,0.0409,0.925,0.911,0.112,0.0354,77.027,1.239743e-07,3
224260,329907.0,0.0704,0.17900,-17.395,0.0409,0.925,0.911,0.112,0.0354,77.027,1.239743e-07,3


In [7]:
df = df.drop_duplicates()

df.shape

(276260, 12)

### Duplicate rows

The dataset contained 1,678 exact duplicate rows. Since these rows contained identical feature values and labels, they were removed to avoid giving repeated observations unnecessary influence during model training.

After removing duplicates, 276,260 observations remain.

---

### Data Types
Check the data types to ensure all features are in a suitable format for analysis and modeling.

In [9]:
df.dtypes

duration (ms)       float64
danceability        float64
energy              float64
loudness            float64
speechiness         float64
acousticness        float64
instrumentalness    float64
liveness            float64
valence             float64
tempo               float64
spec_rate           float64
labels                int64
dtype: object

**Observation:** The target variable was already numerically encoded, and the dataset contained no categorical input features. Therefore, no additional encoding was required.

### Summary Statistics:
Review descriptive statistics to identify unusual values and potential outliers.

In [10]:
df.describe()

,duration (ms),danceability,energy,loudness,speechiness,acousticness,instrumentalness,liveness,valence,tempo,spec_rate,labels
count,2.762600e+05,276260.000000,276260.000000,276260.000000,276260.000000,276260.000000,276260.000000,276260.000000,276260.000000,276260.000000,2.762600e+05,276260.000000
mean,2.324221e+05,0.552509,0.556614,-10.370273,0.087960,0.386906,0.255386,0.189151,0.449543,119.190571,4.758558e-07,1.179049
std,1.170172e+05,0.188941,0.279694,6.672366,0.112613,0.364564,0.373908,0.163513,0.267527,30.473167,9.202471e-07,1.021453
min,6.706000e+03,0.000000,0.000000,-60.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000e+00,0.000000
25%,1.720000e+05,0.431000,0.341000,-12.756000,0.035900,0.033900,0.000000,0.096200,0.220000,95.066750,1.531601e-07,0.000000
50%,2.130530e+05,0.571000,0.591000,-8.406000,0.047100,0.263000,0.001100,0.121000,0.434000,119.936000,2.346238e-07,1.000000
75%,2.648455e+05,0.693000,0.792000,-5.848000,0.082300,0.754000,0.647000,0.226000,0.665000,138.852000,4.454110e-07,2.000000
max,3.919895e+06,0.989000,1.000000,4.882000,0.965000,0.996000,1.000000,1.000000,1.000000,244.947000,5.971860e-05,3.000000


### Long Track Check:
Count how many tracks are longer than 10 minutes to identify potential duration outliers.

In [11]:
(df["duration (ms)"] > 600000).sum()

np.int64(2918)

### Long Track Distribution:
Examine which emotion labels are associated with tracks longer than 10 minutes.

In [12]:
df.loc[df["duration (ms)"] > 600000, "labels"].value_counts()

labels
0    1495
3     752
1     443
2     228
Name: count, dtype: int64

**Observation:** Tracks longer than 10 minutes occur across all emotion classes, with the largest number labeled as "sad". Since the classes differ in size, raw counts alone are not enough to determine whether long duration is particularly associated with a specific emotion.

### Proportion of Long Tracks by Emotion:
Compare the percentage of tracks longer than 10 minutes within each emotion class to account for differences in class size.

In [13]:
long_track_share = (
    df["duration (ms)"].gt(600000)
    .groupby(df["labels"])
    .mean()
    .mul(100)
)

long_track_share

labels
0    1.831479
1    0.419000
2    0.488055
3    1.782497
Name: duration (ms), dtype: float64

Sad: 1,83 %  
Calm: 1,78 %  
Energetic: 0,49 %  
Happy: 0,42 %  

**Observation:** Long tracks are proportionally most common in the sad (1.83%) and calm (1.78%) classes, while they are much less common in happy and energetic tracks. Duration may therefore contain useful information for emotion classification and will be retained for now.

### Extreme Duration Values:
Inspect the longest tracks to determine whether extreme duration values appear plausible before deciding whether any should be removed.

In [14]:
df.nlargest(10, "duration (ms)")[
    ["duration (ms)", "energy", "acousticness", "instrumentalness", "valence", "labels"]
]

,duration (ms),energy,acousticness,instrumentalness,valence,labels
146529,3919895.0,0.0505,0.958,0.9670,0.03590,3
141954,3697969.0,0.0329,0.763,0.8850,0.04100,3
262921,3653565.0,0.1210,0.963,0.9770,0.03490,3
238152,3650064.0,0.0587,0.945,0.9130,0.03320,3
268918,3634235.0,0.0382,0.936,0.9620,0.03710,3
261900,3604214.0,0.0701,0.951,0.9280,0.03360,3
190970,3603000.0,1.0000,0.266,0.6920,0.00001,0
4364,3600000.0,0.9950,0.244,0.9470,0.00729,3
4365,3600000.0,0.8450,0.580,0.0013,0.01240,0
4366,3600000.0,0.9950,0.139,0.9860,0.00001,2


**Decision:** Extreme duration values are retained. Although some tracks are approximately one hour long, their other audio features and emotion labels appear plausible. Without evidence that these observations are erroneous, removing them could discard valid data.

---

### Spec Rate Analysis:
Examine the distribution of `spec_rate` across emotion classes, as its meaning is not documented in the dataset description.

In [15]:
df.groupby("labels")["spec_rate"].describe()

,count,mean,std,min,25%,50%,75%,max
labels,,,,,,,,
0,81628.0,3.604027e-07,1.161205e-06,0.000000e+00,1.177787e-07,1.575613e-07,2.408697e-07,0.000060
1,105728.0,5.891057e-07,8.969227e-07,1.971254e-08,1.785775e-07,2.896122e-07,6.614191e-07,0.000049
2,46716.0,5.020470e-07,6.546406e-07,0.000000e+00,2.060649e-07,3.030976e-07,5.344441e-07,0.000024
3,42188.0,3.864222e-07,6.158038e-07,0.000000e+00,1.698818e-07,2.480247e-07,3.839554e-07,0.000022


**Decision:** `spec_rate` shows some variation across emotion classes, but its meaning is not documented by the dataset creator. Since the feature cannot be reliably interpreted or justified, it will be excluded from the modeling dataset.

In [16]:
df = df.drop(columns=["spec_rate"])

df.columns

Index(['duration (ms)', 'danceability', 'energy', 'loudness', 'speechiness',
       'acousticness', 'instrumentalness', 'liveness', 'valence', 'tempo',
       'labels'],
      dtype='str')

---
### Target Class Distribution:
Check the distribution of emotion classes after data cleaning to identify potential class imbalance before splitting the dataset.

In [17]:
df["labels"].value_counts()

labels
1    105728
0     81628
2     46716
3     42188
Name: count, dtype: int64

happy (1): 105 728  
sad (0): 81 628  
energetic (2): 46 716  
calm (3): 42 188  

**Observation:** The target classes are moderately imbalanced. Happy is the largest class, while calm is the smallest. This imbalance will be considered when splitting the data and evaluating the models.

---

### Features and Target:
Separate the input features (`X`) from the target variable (`y`) before model development.

In [20]:
X = df.drop(columns=["labels"])
y = df["labels"]

print("X shape:", X.shape)
print("y shape:", y.shape)

X shape: (276260, 10)
y shape: (276260,)


### Train-Test Split:
Split the dataset into 80% training data and 20% test data. Stratification is used to preserve the emotion class distribution in both sets.

In [23]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.2,
    random_state=42, # Använd samma slumpmässiga uppdelning varje gång.
    stratify=y
)

print("Training set:", X_train.shape)
print("Test set:", X_test.shape)

Training set: (221008, 10)
Test set: (55252, 10)


### Class Distribution After Split:
Verify that stratification preserved approximately the same emotion class distribution in the training and test sets.

In [24]:
print("Training set:")
print(y_train.value_counts(normalize=True) * 100)

print("\nTest set:")
print(y_test.value_counts(normalize=True) * 100)

Training set:
labels
1    38.271465
0    29.547347
2    16.910248
3    15.270940
Name: proportion, dtype: float64

Test set:
labels
1    38.270108
0    29.548252
2    16.909795
3    15.271845
Name: proportion, dtype: float64


**Observation:** Stratification successfully preserved the emotion class distribution. The proportions are nearly identical in the training and test sets.

---

### Feature Scaling:
Standardize the numerical features so they are on comparable scales. The scaler is fitted only on the training data to prevent data leakage.

In [26]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("Mean:", X_train_scaled.mean())
print("Std:", X_train_scaled.std())

Mean: 5.755509327977863e-17
Std: 1.0


**Observation:** After standardization, the training features have a mean approximately equal to 0 and a standard deviation of 1, confirming that the scaling was successful.  
**Scientific notation:**  
5.755509327977863e-17 = 0.000000000000000057555...

---

### Save Cleaned Dataset:

Save the cleaned dataset for use in the following analysis and modeling notebooks.

In [ ]:
df.to_csv("../data/processed/moodify_clean.csv", index=False)

In [29]:
import os

os.path.exists("../data/processed/moodify_clean.csv")

True

---
## Data Preparation Summary

- No missing values were found in the dataset.
- 1,678 exact duplicate rows were removed, leaving 276,260 observations.
- Extreme track durations were investigated and retained because there was no evidence that they represented erroneous data.
- `spec_rate` was removed because its meaning was not documented and could therefore not be reliably interpreted.
- The dataset contains a moderate class imbalance, with Happy as the largest class and Calm as the smallest.
- An 80/20 stratified train-test split successfully preserved the class distribution.
- Numerical features were standardized using a scaler fitted only on the training data to prevent data leakage.
- The cleaned dataset was saved for use in the following EDA and modeling notebooks.

The dataset is now cleaned and prepared for further analysis and model development.